DataDog/datadog-operator
# ContainerProcessStrategy, allow running non-privileged agents in one container

### What does this PR do?

Feature allowing to configure process strategy within agent containers controlled by `global.containerProcessStrategy.type` property in DatadogAgent. Currently supported `singleProcessContainers` - current behavior and default, `unprivilegedMultiProcessContainer` allowing to run non-privileged agents in one container when privileged Agent isn't needed.

When using `unprivilegedMultiProcessContainer`, if configured feature set doesn't require `system-probe` or `security-agent` Operator will run non-privileged agents in a  single multi-process container. If property is disabled or `system-probe`, `security-agent` required, Operator will create single-process containers.


### Motivation

What inspired you to submit this pull request?

### Additional Notes

How to review?
Key changes in the PR are limited to multiple files. Majority of changes are feature updates, each being self-contained and tests.  
1. Start with CRD change.
2. Next `controllers/datadogagent/feature/types.go`. This defines an interface func `ManageMultiProcessNodeAgent`. This function is implemented by all features to support multi-process agent.
3. `controllers/datadogagent/feature/factory.go` this implements logic for checking `RequiredComponents` and `DatadogAgent` to determine weather to configure default containers (single-process ones) or to use multi-process agent.
4. `controllers/datadogagent/controller_reconcile_agent.go` this decides whether to configure each feature for single or multi-process container and call corresponding `Manage*` function.
5. `global.go`, `default.go` supporting multi-process container.
6. Each feature can be reviewed in isolation.
7. `controllers/datadogagent/controller_v2_test.go` implements controller test to verify how various feature settings determines which containers are created.


### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan
1. Install Operator. 
8. Create manifest not requiring `system-probe` and `security-agent`.
  1. Apply manifest - confirm agent pod is created with 2 or 3 containers. 
  2. Add `global.containerProcessStrategy.type: unprivilegedMultiProcessContainer` to manifest - confirm agents restart with single container.
  9. Update manifest to enabled `cws` feature - confirm agents restart with 4 or 5 containers.
  10. Set `global.containerProcessStrategy.type: singleProcessContainers` - confirm agents restart with same number of containers.
  11. Disable `cws` feature - confirm agents restart with 2 or 3 containers.
  12. Set `global.containerProcessStrategy.type: unprivilegedMultiProcessContainer` - confirm agent restarts with single container.

### Checklist
